core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -371,6 +371,14 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
       throw new SparkException("An application name must be set in your configuration")
     }
 
+    // System property spark.yarn.app.id must be set if user code ran by AM on a YARN cluster
+    // yarn-standalone is deprecated, but still supported
+    if ((master == "yarn-cluster" || master == "yarn-standalone") &&
+        !_conf.contains("spark.yarn.app.id")) {
```

Comment:
@ehnalis the reason I don't think it's necessary is because the user can't get a `spark.yarn.app.id` without using cluster mode. I think it's slightly better to leave it out so we keep the code as simple as possible.